homeassistant/components/statistics/sensor.py

```diff
@@ -1037,31 +1059,66 @@ def _fetch_states_from_database(self) -> list[State]:
             start_date = datetime.fromtimestamp(0, tz=dt_util.UTC)
             _LOGGER.debug("%s: retrieving all records", self.entity_id)
-        return history.state_changes_during_period(
+        # The query will get the list of states in DESCENDING order so that we
+        # can limit the result to self._sample_size. Afterwards reverse the
+        # list so that we get it in the right order again.
+        states = history.state_changes_during_period(
             self.hass,
             start_date,
             entity_id=lower_entity_id,
             descending=True,
             limit=self._samples_max_buffer_size,
             include_start_time_state=False,
         ).get(lower_entity_id, [])
-
-    async def _initialize_from_database(self) -> None:
-        """Initialize the list of states from the database.
-
-        The query will get the list of states in DESCENDING order so that we
-        can limit the result to self._sample_size. Afterwards reverse the
-        list so that we get it in the right order again.
-
-        If MaxAge is provided then query will restrict to entries younger then
-        current datetime - MaxAge.
-        """
-        if states := await get_instance(self.hass).async_add_executor_job(
+        states.reverse()
+        return states
+
+    async def _initialize_from_database(self) -> bool:
+        """Initialize the list of states from the database."""
+        if DATA_RECORDER not in self.hass.config.components:
+            return False
+        states = await get_instance(self.hass).async_add_executor_job(
             self._fetch_states_from_database
+        )
+        if not states:
+            return False
+        for state in states:
+            self._add_state_to_queue(state)
+            self._calculate_state_attributes(state)
+        return True
+
+    async def _initialize_from_restore_state(self) -> bool:
+        """Initialize the list of states from the restore state."""
+        if (last_state := await self.async_get_last_state()) is None:
+            return False
+        if (extra_data := await self.async_get_last_extra_data()) is None:
+            return False
+        try:
+            typed_data = StatisticsSensorExtraStoredData(**extra_data.as_dict())
+        except TypeError:
+            return False
+        self.states = deque(typed_data.states, maxlen=self._samples_max_buffer_size)
+        self.ages = deque(typed_data.ages, maxlen=self._samples_max_buffer_size)
+        self._attr_available = last_state.state != STATE_UNAVAILABLE
+        self._attr_extra_state_attributes[STAT_SOURCE_VALUE_VALID] = (
+            typed_data.source_value_valid
+        )
+        self._attr_native_unit_of_measurement = last_state.attributes.get(
+            ATTR_UNIT_OF_MEASUREMENT
+        )
+        self._attr_device_class = last_state.attributes.get(ATTR_DEVICE_CLASS)
+        self._attr_state_class = last_state.attributes.get(ATTR_STATE_CLASS)
+        return True
```

Comment:
If we add this, should we remove the functionality allowing the statistics sensor to prime its state buffer from the recorder database?

Reply:
yes, I kept the recorder because it's useful for bootstrapping the value when adding a new sensor with a long lookback window. even if it's not 100% correct in that case (due to repeating values), it's still better than nothing